동기화 메모장 [STEP 3] 차차, 예거

Jager-yoo opened the pull request:
안녕하세요~ 코리! 😄 @corykim0829 
동기화 메모장 `스텝3` 구현 완료하여, PR 제출합니다.

외부 저장소를 프로젝트에 적용한 게 처음이라서, 헷갈리는 점도 많고 앱 구조와 같은 `기획적인 부분` 또한 많이 고민해봤어요! 🤔 
앞으로 며칠 안 남았는데, 남은 기간도 잘 부탁드리겠습니다. 🙏🏻 

---

# 🔥 고민했던 점 

### 1️⃣ App단에서 데이터를 관리하기 위한 방법

- Dropbox를 연동함에 따라 데이터를 어느 계층에서 관리해줘야 할지에 대해서 고민했습니다.
기존에는 `MemoSplitViewController`가 가지도록 해줬으나, 데이터는 앱의 전반적인 부분과 관련이 있다고 판단했습니다. 
- 이에 Core Data를 관리하는 `MemoStorage`와 Dropbox를 관리하는 `DropboxManager`를 `AppDelegate`에 위치하도록 변경해주었습니다.
추가적으로 메모를 전반적으로 관리하는 것은 `MemoStorage`이기 때문에, Dropbox와의 연동에 대한 역할도 가져야 한다고 생각하여 내부에 `DropboxManager` 인스턴스를 생성해주었습니다.

<br>

### 2️⃣ Core Data <-> Dropbox 동기화 시점

Core Data를 중심으로 앱의 데이터가 관리되고 있다보니, 어느 시점에 Dropbox에 동기화되어야 하는지 고민했습니다.
최대한 비효율적이고 불필요한 요청을 줄이고자 필요한 상황에만 Dropbox API를 요청하도록 했습니다. 

- Core Data -> Dropbox
    - AppDelegate의 `applicationWillTerminate()`
        - 앱이 종료됨에 따라 비동기 실행을 보장받지 못해 `Thread.sleep(5)`을 하여 실행을 보장해주었습니다. 
    - SceneDelegate의 `sceneDidEnterBackground()`
        - 앱을 백그라운드로 보낼 때에도 Dropbox로 데이터를 보내도록 했습니다.
    - UITextView의 `textViewDidEndEditing()`
        - 텍스트 편집을 마치는 시점에 Core data에서 dropbox로 데이터를 보내도록 했습니다. 
- Dropbox -> Core Data
    - Dropbox 연동 성공시
        - Dropbox에 연동 성공하는 시점에 Dropbox의 최신 데이터를 core data에 동기화합니다.
    - 앱이 실행될 때, Dropbox 연동 정보가 true인 경우
        - `UserDefaults`에 Dropbox 연동 정보를 Bool타입으로 저장하여, 앱이 실행될 때 해당 key에 대한 값이 true인 경우 Dropbox의 데이터를 core data로 받도록 구현했습니다.

앱이 종료되는 경로를 최대한 고민하여 3가지의 경우에 Dropbox로 데이터를 보내도록 했습니다.

<br>

### 3️⃣ 사용자의 선택적 Dropbox 연동

- 앱을 실행할 때 바로 Dropbox 연동 여부를 묻는 것이 아니라, 버튼을 두어 선택적으로 연동할 수 있도록 했습니다. 
- 네트워크가 불가능한 상황에는 core data를 사용해서만 메모를 관리할 수 있도록 해주고, 가능한 경우에만 Dropbox 연동에 따른
동기화를 지원해주었습니다. 
- 또한 연동 성공/실패 여부에 따라 alert를 띄워 사용자에게 연동 성공/실패 여부를 보여주도록 구현했습니다. 

<br>

# 🙋🏻‍♂️ 질문하고 싶은 부분

### 1️⃣ 전반적인 데이터 관리 구조의 변화

- Dropbox 가 추가되기 전에는 `UI` ↔️ `CoreData` 둘 사이에서 메모 데이터가 관리되고 있었는데요!
- Dropbox 이후에는 `Dropbox` ↔️ `CoreData` ↔️ `UI` 이런 식으로 CoreData 를 중심으로 메모 데이터를 관리하고 있습니다.
- 🤚🏻 질문 : 앱 외부에 저장소(클라우드 등)를 두는 경우, `CoreData 를 중심으로` 데이터를 관리하는 로직이 맞는 걸까요?
혹시 개선할 수 있는 부분이 있는지 궁금합니다. ㅎㅎㅎ

<br>

### 2️⃣  willTerminate에서의 Thread.sleep 사용

- 앱이 종료될 때 Core Data의 데이터를 Dropbox에 보내주려고 했습니다. 
- 비동기적으로 작동하다보니 실행을 보장받지 못하여 `Thread.sleep` 을 임의로 5초 부여해 수행 시간을 보장해주었습니다.
- 🤚🏻 질문 : willTerminate같은 부분에서 이렇게 Thread.sleep을 사용하는 것이 바람직한지, 혹시 아니라면 어떤 방식으로 비동기 작업의 완료를 보장해 줄 수 있는지 궁금합니다!

<br>

### 3️⃣ 뷰컨의 required init 에는 fatalError 처리를 해줘도 괜찮을까요?

- 뷰컨의 이니셜라이저를 별도로 정의하는 경우, `required init`까지 구현해야 하는데요.
굳이 완벽하게 작동하도록 만들 필요가 없는 이니셜라이저이지만 `fatalError`를 내부에 사용하는 것은 지양해왔습니다.
- 다만, 뷰컨에 초기화가 필요한 프로퍼티가 있고, 옵셔널로 만들지 않았다면 required init 내부에서도 프로퍼티에 값을 할당해주라는 에러가 뜨더라구요. 😅 
의존성을 줄이기 위해 일부러 `생성자 주입` 방식으로 프로퍼티를 만들어주고 있는데, required init 에서는 생성자 주입이 어려워서, `fatalError`로 처리할지, 아니면 `MemoStorage 타입의 인스턴스를 그냥 생성`해서 넣어줄지 고민하다가, 후자를 선택했습니다.
- 🤚🏻 질문 : 사용하지 않을 `required init` 은 어떻게 처리해주면 좋을까요?
Xcode 가 `fatalError` 을 넣어서 만들어주는 걸 그대로 사용해도 괜찮을까요?
아니면 required init 이 사용되는 경우라도 뷰컨이 제대로 생성될 수 있도록 '의존성 관계'를 조금 왜곡해서라도 구현해두는 게 좋을까요? 🤔 

```swift
// Xcode 가 자동으로 만들어주는 required init
required init?(coder: NSCoder) {
    fatalError("init(coder:) has not been implemented")
}

// 저희가 만든 required init
required init?(coder: NSCoder) {
    self.memoStorage = MemoStorage() // 프로퍼티에 값을 넣어주긴 했는데... fatalError 처리가 깔끔할지...
    super.init(coder: coder)
}
```

<br>

### 4️⃣ 비동기 작업이 중첩되다 보니, 반환값들을 원하는 시점에 받기 어려웠고, 비동기 작업 완료 이후 특정 액션을 위해 결국 NotificationCenter 사용했습니다.

- Dropbox 로그인을 시도 -> 성공하면 동기화 메서드 호출 -> Dropbox 에 저장된 파일명을 비동기로 확인 -> 확인한 파일명으로 파일 데이터를 비동기로 다운로드 -> 파일 다운로드가 완료되면 비동기로 메모를 CoreData 에 생성하며 + tableView.reloadData() 호출
- 아래 코드에서는 이러한 `비동기 작업 중첩`이 이뤄지고 있습니다.
컴플리션 핸들러를 만들어가며 대응했는데요. 마지막에 `reloadData`를 해줄 방법을 계속 찾아보다가, 결국 NotificationCenter 를 사용해서 완성했습니다.
- 🤚🏻 질문 1 : 이렇게 비동기 작업을 중첩하며, `A 가 끝나면 B -> B 가 끝나면 C ...` 와 같은 구조로 로직이 필요한 경우엔 어떻게 코드를 써야 할까요?
저희 로직을 개선해볼 수 있는 힌트를 주실 수 있을까요? 🥺 
- 🤚🏻 질문 2 : `NotificationCenter` 가 나쁜 것은 아니지만, 최대한 지양하라고 들어왔습니다.
코리 또한 NotificationCenter 를 지양하시는 편인가요? 그리고 저희 코드 구조에서는 적절한 사용이었는지도 체크해주시면 좋겠습니다! 🙏🏻 

```swift
// 아래 메서드는 로그인을 시도하는 메서드의 비동기 컴플리션 핸들러에서 호출됩니다. (로그인 성공시 호출)
func synchronizeCoreDataToDropbox() {
    dropboxManager.fetchFilePaths { metaDatas in
        metaDatas.forEach { metaData in
            self.dropboxManager.download(from: "/\(metaData.name)") {
                if self.hasNoMemo(with: $0.id) {
                    self.create(id: $0.id, title: $0.title, body: $0.body, lastModified: $0.clientModified)
                    NotificationCenter.default.post(name: .tableViewNeedUpdate, object: nil)
                }
            }
        }
    }
}
```

ChaminLee committed:
feat: SwiftyDropbox 라이브러리 추가

Jager-yoo committed:
feat: Dropbox 연동 기능 구현

- 테이블뷰의 Navigation Bar 에 Dropbox 와 연동하는 버튼 추가

ChaminLee committed:
feat: DropboxManager 타입 생성

Jager-yoo committed:
refactor: MemoStorage 타입이 기존 SplitVC 에서 AppDelegate 으로 이동

- DropboxManager 타입이 MemoStorage 타입에 종속되게 함

ChaminLee committed:
feat: Dropbox 동기화를 위한 create/update 기능 구현

- dropboxManger내 upload, uploadAll 메서드 구현
- willTerminate, didEnterBackground, didEndEditing에서 core data를 dropbox에 업로드 하도록 구현

Jager-yoo committed:
feat: 새 메모 추가 시, 키보드가 자동으로 올라오는 기능 구현

Jager-yoo committed:
feat: Core Data 메모 삭제 시, Dropbox 에서도 삭제되는 기능 구현

ChaminLee committed:
feat: Dropbox -> CoreData 데이터 불러오기 및 테이블 뷰 업데이트 기능 구현

- dropbox의 path를 찾아 download하여 CoreData 업데이트
- MemoStorage의 create 메서드 파라미터 변경
- DropboxFile 타입 생성
- SplitedText를 extension으로 분리
- Notification.Name을 extension으로 분

Jager-yoo committed:
feat: CoreData 에 이미 존재하는 메모는 Dropbox 에서 다운받지 않는 기능 구현 (중복방지)

- 최초 Dropbox 로그인에 성공하면, UserDefaults 의 연동 정보에 true 값을 넣음
- 앱 실행 시 마다, UserDefaults 를 확인하여, 이미 로그인 기록이 있다면 Dropbox 에서 메모를 다운받아 동기화시킴

ChaminLee committed:
feat: 상수 관리 및 dropbox 연결 alert 추가

Jager-yoo committed:
refactor: Dropbox 메모 업로드/다운로드가 .txt 포맷으로 이뤄지도록 수정

- Navigation Bar 스타일을 Large Title 로 변경
- FileFormat 열거형 추가 생성
- Dropbox upload, delete 메서드 내부에서 id 뒤에 .txt 문자열 붙도록 수정

corykim0829 commented on CloudNotes/CloudNotes/Utility/DropboxManager.swift:
```diff
@@ -0,0 +1,78 @@
+                return
+            }
+            
+            DropboxClientsManager.authorizedClient?.files.upload(path: "/\(id)", mode: .overwrite, input: uploadData)
```
업로드 요청의 성공 여부는 따로 처리하지 않는건가요?

Jager-yoo replied:
저희는 키보드가 내려가거나, 앱이 백그라운드로 갈 때 Dropbox 에 `자동으로 업로드`되는 기능을 넣어놨는데요!
이는 사용자가 의도한 행동이 아니라서, 업로드가 성공했을 때 `Alert`와 같은 알림을 사용자에게 줄 필요는 없다고 생각했습니다.

또한, 업로드에 성공했을 때 받을 수 있는 `response`에는, 저희에게 따로 필요한 정보는 없어보였습니다.
그래서 response 를 따로 처리하지 않고 업로드를 진행해주도록 했습니다!

저희 같은 상황에서, 업로드 성공했을 때 별도의 처리가 필요할까요?

---

우선 error 가 발생했을 때는, 개발자가 어떤 에러인지 확인할 수 있도록 `print` 하는 코드를 추가했습니다! 😄 
이 과정에서 궁금한 게 생겼는데요. ㅎㅎㅎ

실제 배포되는 `상용앱`에서도, error 가 발생했을 때, 개발자가 확인할 수 있는 `print`문을 만들어두시나요??

corykim0829 replied:
프로덕트 코드에는 print문은 사용하지 않고 있어요! 

성공했을 때 Alert가 필요하다는 의견은 아니였고, 유저에게 피드백이 필요한 에러가 발생할 때에는 추가적인 처리가 필요할 것 같아서 여쭤봤었습니다!

corykim0829 commented on CloudNotes/CloudNotes/Utility/DropboxManager.swift:
```diff
@@ -0,0 +1,78 @@
+            return
+        }
+        
+        DropboxClientsManager.authorizedClient?.files.deleteV2(path: "/\(id)", parentRev: nil)
```
삭제 요청의 성공 여부도 따로 처리하지 않는지 궁금합니다!

Jager-yoo replied:
위에 코멘트 드린 것과 유사한 맥락으로, 삭제가 성공했을 때는 사용자에게 별도의 알림은 주지 않고
`error` 가 발생한 경우에만 print 되도록 수정했습니다! 😄

corykim0829 commented on CloudNotes/CloudNotes/Utility/MemoStorage.swift:
```diff
@@ -60,4 +65,40 @@ class MemoStorage {
+        dropboxManager.upload(memos: fetchedMemos)
+    }
+    
+    func synchronizeCoreDataToDropbox() {
```
이 함수의 indent 깊이가 너무 깊어보여요! 정리하면 좋을 것 같아요
또한 함수명을 조금 더 코드가 어떤 일을 하는지 명확하게 명시되면 좋을 것 같아요.

Jager-yoo replied:
의견 감사합니다 코리!

기존에는 여러 역할을 동시에 표현해주려다보니, 메서드 네이밍이 다소 길어지고 불분명하게 된 것 같아요. 😅 
그래서 저희는 좀 더 `본질적인 역할`에 집중해서.. `downloadDropboxData`로 메서드 네이밍을 수정했습니다!

그리고 `indent` 를 좀 더 줄일 수 있는 방법을 고민해봤는데요. 🤔 
그렇게 하려면, 파일 path 를 찾는 메서드와, 다운로드 메서드를 합쳐야만 가능할 것 같아요~!

고민되는 점은... 지금 indent 깊이가 이렇게 된 이유가 2가지 있는데요!
1. 1함수 1기능
2. 비동기 작업의 중첩

이 상황에서 `메서드를 합치지 않고도 indent 를 줄일 방법`이 있을까요? 🥺 
우선 메서드 네이밍만 바꿔서 커밋 올려두겠습니다!

corykim0829 replied:
말씀하신 비동기 작업의 중첩 때문에 indent를 줄이는게 어려워 보이네요.
dropboxManager.download 비동기 응답만 클로저로 작업을 전달을 하면 줄일 수는 있을 것 같아 보입니다만,
시간이 조금 걸릴 것 같아요. 여유가 되시면 하시면 좋을 것 같아요!

corykim0829 commented on CloudNotes/CloudNotes/Utility/MemoStorage.swift:
```diff
@@ -60,4 +65,40 @@ class MemoStorage {
+        }
+    }
+    
+    func checkMemoExistence(with id: UUID) -> Bool {
```
checkMemoExistence라는 함수를 함수 내부를 보지 않고도 추측이 가능하도록 조금 더 명확한 함수명을 사용하면 좋을 것 같아요!

ChaminLee replied:
`hasNoMemo(with id: UUID) -> Bool`로 네이밍 변경했습니다!

corykim0829 reviewed (changes requested):
예거 차차 정말 수고하셨습니다! 👏👏👏

일단 간단하게 앱 테스트를 진행해봤는데, 글을 삭제하면 인덱스 에러가 뜨더라구요! 한번 확인 부탁드립니다 :)

질문에 답변으로도 이야기 했지만, CoreData를 관리하는 객체에서 Dropbox를 처리하는 코드를 추가하기 보다는 각각의 객체를 관리하는 또 하나의 객체가 있어서 이를 하위 객체에게 계속 일을 시키는 형태로 구조를 변경하면 좋을 것 같습니다!

---
### 🙋🏻‍♂️ 질문하고 싶은 부분

### 1️⃣ 전반적인 데이터 관리 구조의 변화

전반적인 코드를 봤을 때에는 MemoStorage 객체를 통해서 CoreData, Dropbox 데이터를 모두 관리하고 있는 것 같아요.
질문에서 `CoreData`를 중심으로 데이터를 관리하는 로직이 맞고 틀리고를 단정짓기는 어렵고, 현재 로직도 좋은데, 하나의 객체가 CoreData, Dropbox를 처리하는 객체에게 각각 일을 처리하도록 하면 더 깔끔한 로직이 될 것 같습니다!

### 2️⃣ willTerminate에서의 Thread.sleep 사용

앱이 종료될 때 sleep을 호출하는 것은 좋은 방법이 아닌 것 같습니다. 앱이 종료되면 클라우드와 같은 곳에 저장되지 않는 건 어쩔 수 없는 부분이라고 생각하면 좋을 것 같아요.

추가적으로 애플 기본 메모장을 살펴보니, 입력 도중 종료해버리면 클라우드에 바로 저장되지는 않지만, 아이폰 메모장에는 계속 저장되는 것 같아요. 고민할 것도 많고 시간이 오래걸리는 작업이니 나중에 고민해보는게 좋을 것 같아요!

### 3️⃣ 뷰컨의 required init 에는 fatalError 처리를 해줘도 괜찮을까요?

fatalError는 최대한 피하는게 좋습니다. 후자로 사용하는게 좋을 것 같아요 :)

### 4️⃣ 비동기 작업이 중첩되다 보니, 반환값들을 원하는 시점에 받기 어려웠고, 비동기 작업 완료 이후 특정 액션을 위해 결국 NotificationCenter 사용했습니다.

NotificationCenter는 앱 내 어디에서든지 받을 수 있다는 점에서, 특정 상황을 위해 사용되는 부분을 보고 NoticiationCenter를 남용한다는 지적이 많습니다. 그래서 지양하는 분들이 있는 걸로 알고 있어요!

현재 MemoSplitViewController에 MemoStorage 인스턴스를 선언했고, 이 인스턴스의 함수에서 비동기로 결과를 받아와서 MemoSplitViewController에서 처리를 해도 충분히 구현은 가능한 것 같아요.

이번 구현에서는 NotificationCenter을 사용하는게 조금은 과한 구현일 수 도 있다고는 생각이 듭니다!

ChaminLee committed:
refactor: MemoStorage -> CoreDataManager 네이밍 변경 및 프로토콜 분리

Jager-yoo committed:
refactor: CoreDataManager, DropboxManager 를 관리하는 상위 객체인 MemoStorage 타입 생성

- applicationWillTerminate 메서드 제거

ChaminLee committed:
refactor: Dropbox API 에러 처리 로직 추가

Jager-yoo committed:
refactor: NotificationCenter 사용 로직 제거

- Dropbox 와 동기화가 성공했을 때, NotificationCenter 가 아닌, Completion Handler 를 통해 Alert 를 띄우고 MemoTableVC 의 화면을 리로드하도록 수정함

ChaminLee committed:
feat: Dropbox login/logout 버튼 추가

Jager-yoo committed:
refactor: Dropbox 동기화 메서드 네이밍 변경

- synchronizeCoreDataToDropbox -> downloadDropboxData

Jager-yoo commented:
## 🛠️ 스텝3 리팩토링 내용

### 1️⃣ CoreData, Dropbox 데이터를 모두 관리하는 상위 객체 생성

- CoreDataManager, DropboxManager 타입을 관리하는 상위 객체인 MemoStorage 를 구현하고 프로토콜도 추가 생성했습니다.

### 2️⃣ applicationWillTerminate 메서드 제거

- 말씀해주신 아이폰 메모장처럼, 저희 앱 또한 CoreData 에는 글자 단위로 저장이 되고 있기 때문에, 중간에 강제 종료가 되더라도 동기화에는 문제가 없었습니다. 😄 

### 3️⃣ NotificationCenter 제거

- 말씀해주신 힌트를 적용하여 NotificationCenter 를 없앨 수 있었습니다. 감사합니다! 👍🏻

ChaminLee committed:
fix: 메모 삭제 및 동기화 버그 개선

- textViewEndEditing의 메모 삭제 로직이 에러를 불러일으킬 요소가 많아 제거
- upload시에 유효한 indexPath 인지 확인 후 진행하도록 개선

Jager-yoo committed:
refactor: 빈 메모 중복 생성을 방지하는 기능 구현

- Navigation Bar 에 위치한 + 버튼의 isEnabled State 를 조정하는 식으로 사용자가 빈 메모를 중복해서 추가하지 못하도록 함

ChaminLee committed:
feat: tableView가 editing될 때 selection이 유지되도록 구현

Jager-yoo committed:
fix: 화면 회전 시, UIActivityViewController 위치가 항상 중앙에 오도록 수정

Jager-yoo committed:
fix: 빈 메모 생성하고 바로 삭제 시, 메모 추가 버튼이 다시 활성화되도록 수정
